src/osd/PrimaryLogPG.cc

```diff
@@ -8207,6 +8626,105 @@ void PrimaryLogPG::process_copy_chunk(hobject_t oid, ceph_tid_t tid, int r)
   kick_object_context_blocked(cobc);
 }
 
+void PrimaryLogPG::process_copy_chunk_manifest(hobject_t oid, ceph_tid_t tid, int r, uint64_t offset)
+{
+  dout(10) << __func__ << " " << oid << " tid " << tid
+	   << " " << cpp_strerror(r) << dendl;
+  map<hobject_t,CopyOpRef>::iterator p = copy_ops.find(oid);
+  if (p == copy_ops.end()) {
+    dout(10) << __func__ << " no copy_op found" << dendl;
+    return;
+  }
+  CopyOpRef ori_cop = p->second;
+  CopyOpRef cop = ori_cop->chunk_cops[offset];
+
+  if (tid != cop->objecter_tid) {
+    dout(10) << __func__ << " tid " << tid << " != cop " << cop
+	     << " tid " << cop->objecter_tid << dendl;
+    return;
+  }
+
+  if (cop->omap_data.length() || cop->omap_header.length()) {
+    r = -EOPNOTSUPP;
+  }
+
+  cop->objecter_tid = 0;
+  cop->objecter_tid2 = 0;  // assume this ordered before us (if it happened)
+  ObjectContextRef& cobc = ori_cop->obc;
+
+  if (r < 0) {
+    ori_cop->failed = true;
+    goto out;
+  }   
+
+  if (ori_cop->failed) {
+    return;
+  }                                                                                                                  
+  assert(cop->rval >= 0);
+
+  if (ori_cop->num_chunk) {
```

Comment:
I think we'll want to do something a bit more clever here.  As it stands, if you have a 4mb object with 100 40k chunk, we'll end up generated 100 repops to promote it.  Instead, I think we want to read some portion of the object (say, 4 mb at a time), accumulate it all in memory, and then generate a single repop for that section before moving on to the rest of the object.  This would be similar to what recovery is doing (i think we default to 4mb chunks there?).

Reply:
Right. I will fix it as you comment.